rust/src/push/evaluator.rs

```diff
@@ -395,59 +422,45 @@ impl PushRuleEvaluator {
     /// Evaluates a `related_event_match` condition. (MSC3664)
     fn match_related_event_match(
         &self,
-        event_match: &RelatedEventMatchCondition,
-        user_id: Option<&str>,
+        rel_type: &str,
+        include_fallbacks: Option<bool>,
+        key: Option<Cow<str>>,
+        pattern: Option<Cow<str>>,
     ) -> Result<bool, Error> {
         // First check if related event matching is enabled...
         if !self.related_event_match_enabled {
             return Ok(false);
         }
 
         // get the related event, fail if there is none.
-        let event = if let Some(event) = self.related_events_flattened.get(&*event_match.rel_type) {
+        let event = if let Some(event) = self.related_events_flattened.get(rel_type) {
             event
         } else {
             return Ok(false);
         };
 
         // If we are not matching fallbacks, don't match if our special key indicating this is a
         // fallback relation is not present.
-        if !event_match.include_fallbacks.unwrap_or(false)
-            && event.contains_key("im.vector.is_falling_back")
-        {
+        if !include_fallbacks.unwrap_or(false) && event.contains_key("im.vector.is_falling_back") {
             return Ok(false);
         }
 
         // if we have no key, accept the event as matching, if it existed without matching any
         // fields.
-        let key = if let Some(key) = &event_match.key {
+        let key = if let Some(key) = key {
```

Comment:
I wonder if here down to +463 might be clearer like this? :shrug:

```rust
        match (key, pattern) {
            (None, _) => Ok(true),
            (Some(key), None) => Ok(false),
            (Some(key), Some(pattern) => self.match_event_match(event, &key, &pattern),
        }
```

Reply:
That might indeed be a bit clearer, strays a bit from the current logic, but I think that's OK.

Reply:
FWIW that suggestion wasn't meant to introduce any logical changes---what difference in logic do you see here?

Reply:
s/logical/logic or s/logical/behaviour

Reply:
Much simpler indeed!